### DISCUSSION TITLE
What setting causes the cursor to jump into the next pair rather than create a new pair?

### DISCUSSION THREAD

**Discussion by pooriajr** [LunarWatcher/auto-pairs#40] [Q&A] [answered]
Sorry if this is a dumb question, but I've been struggling with this constantly. Here's the situation:
The cursor is in front of a pair, separated by a space:
`| "foo"`
I want to press `"` to open a new pair of quotes like so:
`"|" "foo"`
But instead, if I press `"` I get this:
`"|foo"`

Is there a way to disable this behavior? Thank you for maintaining this great plugin

**Comment by LunarWatcher** [comment, marked as answer]
Sorry for the late response - this managed to slip through my stack of stuff to do (been getting a lot of emails for a couple days, and this managed to slip my mind. My bad).

The option for this is `g:AutoPairsSearchCloseAfterSpace`. That said, there's an underlying bug here - with it disabled, it outputs `" "foo"`. This is a symptom of the actual problem; for whatever reason, the plugin is registering an imbalance that isn't there. That's probably why it's jumping too, but I'm not entirely sure. The jump logic for when open == close relies on simply guessing what's an open quote and what's a close quote, and what's caused by an imbalance. Could probably be improved, but this falls into the category of stuff that has edge-cases when not doing contextual parsing.

Not sure how much of the processing is bad here, so I'm gonna open an issue to track this as well

---

This has now been fixed, as of #54. Feel free to open a new bug report (preferably as an issue) if anything now breaks.

**Reply by LunarWatcher** [reply]
I've identified the problem, but implementing it is tricky, and borders a 4.0.0-change. I'll have to push it as a 3.0.0-bugfix, but I digress.

This is a bug [partly inherited from jiangmiao](https://github.com/jiangmiao/auto-pairs/blob/master/plugin/auto-pairs.vim#L224-L226), but it wasn't a problem for two major reasons:
1. No "proper" balance checking
2. No optional space checking (and I believe `| "foo"`, " at | would force a jump) 

 I can fix it, but it requires moving the jump logic to its own function, and doing (another) rewrite of the core insert logic, and I starting that this late during the day is asking for trouble. Got no timeline on this either, because I have no idea how bad stuff will break when I start the rewrite. There's also no cheap hacks I can use here

**Reply by pooriajr** [reply]
Thanks for looking into this. Actually I had the same issue very frequently using jiangmiao's plugin. I'm confused why it doesn't seem to be discussed by anyone - it happens quite often for me and destroys my typing flow. Anyway, thanks for for the help, I'll just live with it for now. 👍

**Reply by LunarWatcher** [reply]
Just posting an update because it's been a couple months -- my schedule has started clearing up, so I'm on top of this now. Not sure how I'm gonna go about anything yet, so no guarantees on when I'll be done, but I estimate a couple weeks, possibly more if stuff:tm: takes up a lot of my time again. Gonna need to find a way to hotpatch it, and I have some work that I never finished and now don't remember or understand what does, but it's on the develop branch for tracking

**Reply by LunarWatcher** [reply]
Just merged #54 -- stuff did get in the way (long couple months, but 3 weeks and 2 days isn't too far above my estimate) but I've fixed it. The patch here was to hook in the balance checker for a contextual break; essentially, if there's an odd number of quotes (or other open == close, I believe) after the cursor, and all the old conditions hold, it skips to close checking. The addition of balance checking appears to be what made this a noticable problem. Anyway, it's all gone now.

This doesn't appear to interfere with the balance checking, which is great, but is enough to make sure closure isn't done unless there's an odd number after the cursor. It's still able to balance `|" "foo"`, " at | to `"|" "foo"`, because of the double balance check, I assume. It's 2 in the morning and I don't feel like checking if this assumption is correct.

Feel free to open an issue if this causes new bugs though (or if you find other, unrelated bugs, obviously) - all the tests currently pass, so while I don't believe it should be a problem, you never know when old code comes back to haunt you :')

**Reply by pooriajr** [reply]
Woohooo! It works perfectly! Thank you, you're awesome! 😄
